### DISCUSSION TITLE
fix event sequence issues between user_solution_commit and provider_approve

### DISCUSSION THREAD

**Discussion by goastler** [prosopo/protocol#170] [Ideas]

**Comment by goastler** [comment]
In short, a user submits a commitment to the contract and the provider and awaits the provider to approve or disapprove. However, between these two events (submit and approve/disapprove) there can be many other events occur which can disrupt the captcha process overall. For instance, after a user submits but before an approve/disapprove from a provider, the following could occur:
1. provider deregister/unstake
2. provider suspended
3. provider payee change
4. provider fee change
5. provider balance changes below 
6. provider service_origin change
7. provider dataset id / content id change
8. dapp deregisters
9. dapp becomes non-active
10. dapp balance goes below threshold (stake_default)
11. dapp owner change
12. dapp client_origin change
13. governance status for the contract changed, e.g. suspended
etc

We need to consider these occurrences during the captcha process, and adjust the contract to handle any of these occurrences.

**Comment by goastler** [comment]
deregister of provider needs to be changed to a request basis

**Comment by forgetso** [comment]
### provider deregister/unstake

Providers should move into the `Suspended` state prior to de-registering. Whist suspended, Providers will be able to approve or disapprove commitments but will not be selected as random `Active` Providers. Alternatively, if `Suspended` means something else (like suspended due to bad behavioud), an additional state should be created to cover the period after requesting to deregister, before deregistering.

### provider suspended

???

### provider payee change

This is not a problem as change will only affect a minimal number of commitments. Provider selection criteria should be implemented to make sure that Providers do not become too expensive for Dapps.

### provider fee change

This is not a problem as change will only affect a minimal number of commitments. Provider selection criteria should be implemented to make sure that Providers do not become too expensive for Dapps.

### provider balance changes below provider_stake_default

?? Provider will not be able to approve commitment

### provider service_origin change

Doesn't matter. User has already contacted Provider and received captcha.

### provider dataset id / content id change

Doesn't matter. Commitment has already been made by user. Provider has data to build commitment ID locally and retrieve commitment from contract.

### dapp deregisters

Doesn't matter.

### dapp becomes non-active

Doesn't matter.

### dapp balance goes below threshold (stake_default)

This is checked before provider approves or disapproves. https://github.com/prosopo/protocol/blob/d5fc185c89bfaeb90eaed567180200bd12db4abe/contracts/lib.rs#L887

### dapp owner change

Doesn't matter.

### dapp client_origin change

Client origin being removed.

### governance status for the contract changed, e.g. suspended

Contract governance status being removed.
